# <center> Homework 1
<center> Luke Villanueva \\ MATH151A \\ 4/11/23

# Table of Contents
---
* [Problem 1](#problem-1)
* [Problem 2](#problem-2)
* [Problem 3](#problem-3)
* [Problem 4](#problem-4)
* [Problem 5](#problem-5)
* [Coding Part](#coding-part)

In [61]:
# imported libraries
import numpy as np
from icecream import ic
import sympy as sym

# Problem 1
## a.
The Intermediate Value Theorem says that given a continuous function and an interval, then if you give me any number, I can plug that number into the function and I can give you a result that will always be bounded within a certain distance. That said distance will be the interval plugged into the function. For example, if $f(1) = 2$ and $f(3) = 4$, then there is some value $c$ between 1 and 3 such that $f(c) = 3$.

## b.

In [62]:
# plug in interval into function
x = 0.2 
y = 0.3 

# equation
def func(x):
    return x * np.cos(x) - 2 * x**2 + 3 * x - 1

# get results of intervals
print(func(x))
print(func(y))

-0.28398668443175157
0.0066009467376817454


And so, based on these results, because the left bound of the range is negative and the right is positive, we can use the IVT to conclue that there should exist a value $c$ such that $f(c) = 0$. And so, there is a solution to the given non-linear equation.

# Problem 2

Taylor Formula: $\sum_{n=0}^{\infty} \frac{f^{(n)}}{n!}(x-a)^{n}$ 

In [63]:
x = sym.symbols("x")
sym.init_printing()
f1 = sym.sqrt(x+1)
f2 = sym.diff(f1)
f3 = sym.diff(f2)
f4 = sym.diff(f3)
f1

  _______
╲╱ x + 1 

In [64]:
f2

     1     
───────────
    _______
2⋅╲╱ x + 1 

In [65]:
f3

    -1      
────────────
         3/2
4⋅(x + 1)   

In [66]:
f4

     3      
────────────
         5/2
8⋅(x + 1)   

And so, the function's derivatives are:

$f(x) = \sqrt{x + 1}$  
$f'(x) = \frac{1}{2\sqrt{(x+1)}}$  
$f''(x) = \frac{-1}{4(x+1)^{\frac{3}{2}}}$  
$f'''(x) = \frac{3}{8(x+1)^{\frac{5}{2}}}$ 

If $a = 0$, the evaluated Taylor polynomial to when $n = 3$ is,

$P_{3}(0) = 1 + \frac{x}{2} - \frac{x^{2}}{8} + \frac{x^{3}}{16}$

In [67]:
# making function to approximate any x
def taylor_poly(x):
    return 1 + x/2 - (x**2)/8 + (x**3)/16

# approximating sqrt(0.5) -> x = -0.5
taylor_poly(-0.5)

0.7109375

In [68]:
# approximating sqrt(0.75) -> x = -0.25
taylor_poly(-0.25)

0.8662109375

In [69]:
# approximating sqrt(1.25) -> x = 0.25
taylor_poly(0.25)

1.1181640625

In [70]:
# approximating sqrt(1.5) -> x = 0.5
taylor_poly(0.5)

1.2265625

In [71]:
# calculate absolute error of sqrt(0.5)
abs(np.sqrt(0.5) - taylor_poly(-0.5))

0.0038307188134524273

In [72]:
# calculate absolute error of sqrt(0.75)
abs(np.sqrt(0.75) - taylor_poly(-0.25))

0.0001855337155614034

In [73]:
# calculate absolute error of sqrt(1.25)
abs(np.sqrt(1.25) - taylor_poly(0.25))

0.00013007375010509747

In [74]:
# calculate absolute error of sqrt(1.5)
abs(np.sqrt(1.5) - taylor_poly(0.5))

0.0018176286084110593

# Problem 3
## a.

\begin{gather*}
0.101 * 10 (e^{1.53})^{4} - 0.462 * 10 (e^{1.53})^{3} - 0.311 * 10 (e^{1.53})^{2} + 0.122 * 10^{2} e^{1.53} - 0.199 * 10 \newline
\implies
\newline
0.101 * 10 (4.62)^{4} - 0.462 * 10 (4.62)^{3} - 0.311 * 10 (4.62)^{2} + 0.122 * 10^{2} * 4.62 - 0.199 * 10 
\newline
\implies
\newline
0.101 * 10 * fl_{r}(455.5834) - 0.462 * 10 * fl_{r}(98.6111) - 0.311 * 10 * fl_{r}(21.3444) + 0.122 * 10^{2} * 0.462 * 10 - 0.199 * 10
\newline
\implies
\newline
0.101 * 10 * fl_{r}(0.4555 * 10^{3}) - 0.462 * 10 * fl_{r}(0.9861 * 10^{2}) - 0.311 * 10 * fl_{r}(0.2134 * 10^{2}) + 0.122 * 10^{2} * 0.462 * 10 - 0.199 * 10
\newline
\implies
\newline
0.101 * 10 * 0.456 * 10^{3} - 0.462 * 10 * 0.986 * 10^{2} - 0.311 * 10 * 0.213 * 10^{2} + 0.122 * 10^{2} * 0.462 * 10 - 0.199 * 10
\newline
\implies
\newline
0.461 * 10^{3} - 0.456 * 10^{3} - 0.662 * 10^{2} + 0.564 * 10^{2} - 0.199 * 10
\newline
\implies
\newline
0.5 * 10 - 0.98 * 10 - 0.199 * 10
\newline
\implies
\newline
-0.679 * 10
\newline
\implies
\newline
-6.79
\end{gather*}

## b. 

\begin{gather*}
f(x) = -1.99 + e^{x}(12.2 + e^{x}(-3.11 + e^{x}(-4.62 + 1.01e^{x})))
\end{gather*}

## c. 

\begin{gather*}
-1.99 + e^{1.53}(12.2 + e^{1.53}(-3.11 + e^{1.53}(-4.62 + 1.01e^{1.53})))
\newline
\implies
\newline
-1.99 + 4.62(12.2 + 4.62(-3.11 + 4.62(-4.62 + 1.01*4.62)))
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (0.122 * 10^{2} + 0.462 * 10(-0.311 * 10 + 0.462 * 10(-0.462 * 10 + 0.101 * 10 * 0.462 * 10)))
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (0.122 * 10^{2} + 0.462 * 10(-0.311 * 10 + 0.462 * 10(-0.462 * 10 + 0.467 * 10)))
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (0.122 * 10^{2} + 0.462 * 10(-0.311 * 10 + 0.462 * 10(0.5 * 10^{-1})))
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (0.122 * 10^{2} + 0.462 * 10(-3.11 + 0.231))
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (0.122 * 10^{2} + 0.462 * 10(-2.879))
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (0.122 * 10^{2} + 0.462 * 10(-0.288 * 10))
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (0.122 * 10^{2} - 0.133 * 10^{2})
\newline
\implies
\newline
-0.199 * 10 + 0.462 * 10 (-0.11 * 10)
\newline
\implies
\newline
-0.199 * 10 + -0.508 * 10
\newline
\implies
\newline
-0.199 * 10 + -0.508 * 10
\newline
\implies
\newline
-0.707 * 10
\newline
\implies
\newline
-7.07
\end{gather*}


## d. 

To get better accuracy, we need the relative error percentage:

In [75]:
# non nested error percentage
abs(-6.79 - -7.61)/abs(-7.61)

0.10775295663600529

In [76]:
# nested error percentage
abs(-7.07 - -7.61)/abs(-7.61)

0.07095926412614981

The nested form had a 7.1% error while the non-nested form had a 10.8% error. This means that the nested form is more accurate.

# Problem 4

$\frac{1}{3}x^{2} - \frac{123}{4}x + \frac{1}{6} = 0$

## a. 

$\frac{b - \sqrt{b^2 - 4ac}}{2a}$

\begin{gather*}
\frac{-\frac{123}{4} - \sqrt{(-\frac{123}{4})^{2} - 4(\frac{1}{3})(\frac{1}{6})}}{2(\frac{1}{3})}
\newline
\implies
\newline
\frac{-0.3075 * 10^{2} - \sqrt{(-0.3075 * 10^{2})^{2} - 4(0.3333)(0.1666)}}{0.6666}
\newline
\implies
\newline
\frac{-0.3075 * 10^{2} - \sqrt{(-0.3075 * 10^{2})^{2} - 4(0.5552 * 10)}}{0.6666}
\newline
\implies
\newline
\frac{-0.3075 * 10^{2} - \sqrt{(-0.3075 * 10^{2})^{2} - 0.2220}}{0.6666}
\newline
\implies
\newline
\frac{-0.3075 * 10^{2} - \sqrt{0.9455 * 10^{3} - 0.2220}}{0.6666}
\newline
\implies
\newline
\frac{-0.3075 * 10^{2} - \sqrt{945.5 - 0.2220}}{0.6666}
\newline
\implies
\newline
\frac{-0.3075 * 10^{2} - \sqrt{945.2}}{0.6666}
\newline
\implies
\newline
\frac{-30.75 - 30.74}{0.6666}
\newline
\implies
\newline
\frac{-30.75 - 30.74}{0.6666}
\newline
\implies
\newline
\frac{-61.49}{0.6666}
\newline
\implies
\newline
-92.24
\end{gather*}

## b.

\begin{gather*}
x_{\text{true}} = 0.0054 \\ 
x_{1} = -92.24 \\
\text{Absolute error: } \implies |-92.24 - 0.0054| = 92.2454 \\
\text{Relative error: } \implies \frac{|-92.24 - 0.0054|}{|0.0054|} = 17082.4815 \implies 1,708,248.15\% \text{ error}
\end{gather*}

## c.

$\frac{b - \sqrt{b^2 - 4ac}}{2a} \implies \frac{-4ac}{2a(b + \sqrt{b^{2} - 4ac})}$

\begin{gather*}
\frac{-4ac}{2a(b + \sqrt{b^{2} - 4ac})} \\ 
\implies \\
\frac{-0.2220}{0.6666(-30.75 + \sqrt{945.5 - 0.2220})} \\
\implies \\
\frac{-0.2220}{0.6666(-30.75 + \sqrt{945.2})} \\
\implies \\
\frac{-0.2220}{0.6666(-30.75 - 30.74)} \\
\implies \\
\frac{-0.2220}{0.6666(-61.49)} \\
\implies \\
\frac{-0.2220}{-40.98} \\
\implies \\
0.005417
\end{gather*}

## d. 

\begin{gather*}
x_{\text{true}} = 0.0054 \\ 
x_{1} = 0.005417 \\
\text{Absolute error: } \implies |0.005417 - 0.0054| = 0.000017 \\
\text{Relative error: } \implies \frac{|0.005417 - 0.0054|}{|0.0054|} = 0.003148 \implies 0.3148\% \text{ error} \\
\end{gather*}

## e. 

The source of increasing error percentages would be the number of computations that has to be done to get the result. The more computations done, the more inaccuracies will build up because of the chopping process.

# Problem 5

## a. 

### i. 

\begin{gather*}
\frac{4}{5} + \frac{1}{3} = 0.8 + 0.3\bar{3} = 1.13\bar{3}
\end{gather*}

### ii.

\begin{gather*}
(\frac{1}{3} + \frac{3}{11}) - \frac{3}{20}= 0.3\bar{3} + 0.27\bar{27} - 0.15 = 0.4560\bar{60}
\end{gather*}

## b.

### i.

\begin{gather*}
\frac{4}{5} + \frac{1}{3} = 0.8 + 0.333 = 1.133
\end{gather*}

### ii.

\begin{gather*}
(\frac{1}{3} + \frac{3}{11}) - \frac{3}{20}= 0.333 + 0.272 - 0.15 = 0.455
\end{gather*}

## c.

### i.

\begin{gather*}
\frac{4}{5} + \frac{1}{3} = 0.8 + 0.333 = 1.133
\end{gather*}

### ii.

\begin{gather*}
(\frac{1}{3} + \frac{3}{11}) - \frac{3}{20}= 0.333 + 0.273 - 0.15 = 0.456
\end{gather*}

## d.

### i.

\begin{gather*}
\frac{|1.133 - 1.133|}{|1.133|} = 0\% \text{ error}
\end{gather*}

### ii.

\begin{gather*}
\frac{|0.456 - 0.455|}{|0.455|} = 0.002197 \implies 0.2197\% \text{ error}
\end{gather*}

# Coding Part

In [91]:
# condition to stop loop: reach counter number or until f(midpoint) = 0
# default loop amount is 100
def bisec(func, a, b, c = 100):
    i = 0
    mid = (a + b) / 2
    while(i != c and func(mid) != 0):

        mid = (a + b) / 2

        # if func(mid) is negative and func(a) is negative OR 
        # func(mid) is + and func(a) is +, then [mid,b]

        # if f(mid) and f(a) have same sign,
        if((func(mid) < 0 and func(a) < 0) or (func(mid) > 0 and func(a) > 0)):
            a = mid
        # else, f(mid) and f(b) have same sign
        else:
            b = mid

        # increment i
        i += 1
    
    # print iteration amount
    if(i == c):
        print("Program ran max iterations:", i)
    else:
        print("Program ran", i, "iterations")
    # result
    return mid

# this function is defined to have sqrt(3) as a root, so use
# bisec function on this function
def func(x):
    return x**2 - 3

# result
print(bisec(func,0,4,16))

# proof of 10^-5 tolerance
print(bisec(func,0,4,16) - np.sqrt(3))

Program ran max iterations: 16
1.73199462890625
Program ran max iterations: 16
-5.6178662627193177e-05


It took approximately 13 iterations for the resulting root value to have at least a $10^{-5}$ tolerance amount. 